### DISCUSSION TITLE
Add UnhandledInput singleton

### DISCUSSION THREAD

**Discussion by markushonok** [godotengine/godot-proposals#9189] [Scripting]
Add an UnhandledInput singleton with a design symmetric to the Input singleton. 

While writing your own shouldn't be a problem (I think), there are some reasons why this should be built into Godot Engine:
- Non-symmetry due to the already existing Input singleton and the lack of an unhandled input counterpart;
- Inconvenience for beginners, who would be much more comfortable writing simple code using `Input`/`UnhandledInput` singletons;
- Will be heavily used when writing game controls;
- Such a small thing for an engine user should not lead to writing boilerplate code*, install extensions or misuse the Input singleton with subsequent UI problems.

*boilerplate code example (movement without physics for simplicity):
```gdscript
var _direction = Vector2.zero

func _unhandled_input(event):
    if event.is_action_pressed("move_left", true):
        _direction.X = -1
    elif event.is_action_pressed("move_right", true):
        _direction.X = 1
    else _direction.X = 0

    if event.is_action_pressed("move_forward", true):
        _direction.Y = -1
    elif event.is_action_pressed("move_back", true):
        _direction.Y = 1
    else _direction.Y = 0

func _process(delta):
    position += direction * delta
```

Code with UnhandledInput singleton:
```gdscript
func _process(delta):
    var direction = UnhanledInput.get_vector("move_left", "move_right", "move_forward", "move_back")
    position += direction * delta
```

Most guides for creating 2d character movement use the `Input` singleton. Because of this, input handling is not taken into account and this can lead to problems with the UI in the future. This speaks to the convenience of the `Input` singleton and the lack of an unhandled alternative.

There are additional methods in that singleton, such as `get_axis()`, `get_vector()`, `is_action_just_*()`and others. Using these convenient methods makes the code much shorter and simpler.

**Comment by Sauermann** [comment]
Can you please provide more details:
- Which of the events should be accessible via the suggested UnhandledInput singleton?
- How would you want your script to look like after the introduction of this new singleton?

**Comment by markushonok** [comment]
> Can you please provide more details:
> 
> * Which of the events should be accessible via the suggested UnhandledInput singleton?
> * How would you want your script to look like after the introduction of this new singleton?

Now I have expanded the description of my proposal. I believe that all methods from `Input` singleton should be added to the `UnhandledInput` singleton.

**Comment by AThousandShips** [comment]
This wouldn't work without breaking compatibility, as `Input` [isn't affected](https://docs.godotengine.org/en/stable/classes/class_control.html#class-control-method-accept-event) by the handled/unhandled features which are specific to `Viewport`s/`Control`s. So for this to make sense `Input` would have to now respect this or `UnhandledInput` wouldn't be a complement to `Input` which would be confusing, further the handled/unhandled state isn't global, it's specific to viewports and isn't stored in the event at all

So:
* How would `UnhandledInput` work? Where would it get the information?
* How would `Input` change, if at all, with this? And same question as above.
* How would `get_axis` and `get_vector` work? They relate to multiple events, some possibly in the past, how would you handle part of the event being handled and part of it not? How would this update each frame? How would it handle different event sources?
* How would actions be handled? Actions aren't part of this system *either* in this sense, as action based events are not sent by default, only the underlying events triggering them. How would the fact that multiple actual input can map to the same action, or that the same input can map to multiple actions? When is an action unhandled? When is it not?

I don't think this is possible to implement due to the way input management is done, and the fact that the handled/unhandled part isn't global but local, and only deals with raw input events
